### ISSUE TITLE
Replace cJSON with another JSON library

### ISSUE THREAD

**Issue by tristan957** [hse-project/hse#580]
### Description

Problems with cJSON:

1. Undermaintained
2. "Lots" of allocations (how big of an issue is this? not sure)
3. Will leak memory depending on which cJSON APIs you use

Alternatives:

1. [yyjson](https://github.com/ibireme/yyjson)
2. [jansson](https://github.com/akheron/jansson)
3. Wrap a C++ JSON library to create a C library

cc @alexttx

**Comment by tristan957** [comment]
A C++ library to wrap could be https://github.com/nlohmann/json. It is extremely popular

**Comment by beckerg** [comment]
"We used all the operator magic of modern C++ to achieve the same feeling in your code."

That's kinda frightening!  Why not yyjson?

**Comment by tristan957** [comment]
We can use it. I am just giving a wide array of options.

**Comment by alexttx** [comment]
| library | stars | watchers | forks |
|---|---|---|---|
| yyjson |  2.3k | 54 | 220 | 
| jansson | 2.7k | 125 | 772 |
| nlohmann/json | 33.2k | 756 | 5.7k |

**Comment by tristan957** [comment]
One thing Greg had suggested was creating like a libhse-json that was just a wrapper around whatever library we choose, so that if we ever had to swap it out, it would be pretty straightforward to do in the future. Technically that work could happen right now wrapping cJSON.

But I am not sure how much gain there is in that approach.

**Comment by tristan957** [comment]
Last release of jansson: 9/9/2021
Last release of yyjson: 12/12/2022
Last release of nlohmann/json: 8/12/2022

**Comment by beckerg** [comment]
jansson looks decent, but i didn't really dig in.  the extra maturity could be a good thing...

**Comment by tristan957** [comment]
I edited Alex's comment with the table to include the cJSON stats. Technically cJSON is "mature", but we have a lot of problems with it that can't be fixed because it is unmaintained.

The one thing nlohmann/json has going for it is that it has at least one corporate sponsor.

**Comment by beckerg** [comment]
cJSON may be mature, but the using doubles for integers stupidity is broken by design.  So we need one that has actual support for 64-bit integers and is reasonably easy to use.  yyjson looks pretty C-ish, jansson looks like you need to learn a bunch of magic printf-like escapes.  But I haven't dug in much, would need to see some examples in-situ...

**Comment by alexttx** [comment]
One advantage of a wrapper is the ability to test the different libraries for performance, memory usage, etc.  If the underlying APIs are too different, the wrapper may be more trouble than its worth.  It might also be forced into adding its own overhead.

**Comment by alexttx** [comment]
So the only way to instantiate a JSON object in `yyjson` to parse a string?  Does that work for HSE?

**Comment by tristan957** [comment]
No. When you parse a file you get a different struct type in yyjson than you do if you are building a json object by hand.
